streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -38,346 +40,626 @@
+
+    void setMainConsumer(final Consumer<byte[], byte[]> consumer) {
+        this.mainConsumer = consumer;
+    }
+
+    public StoreChangelogReader(final StreamsConfig config,
+                                final LogContext logContext,
+                                final Consumer<byte[], byte[]> restoreConsumer,
+                                final StateRestoreListener stateRestoreListener) {
+        this.log = logContext.logger(StoreChangelogReader.class);
+        this.state = ChangelogReaderState.ACTIVE_RESTORING;
         this.restoreConsumer = restoreConsumer;
-        this.pollTime = pollTime;
-        this.log = logContext.logger(getClass());
-        this.userStateRestoreListener = userStateRestoreListener;
+        this.stateRestoreListener = stateRestoreListener;
+
+        // NOTE for restoring active and updating standby we may prefer different poll time
+        // in order to make sure we call the main consumer#poll in time.
+        // TODO: once both of these are moved to a separate thread this may no longer be a concern
+        this.pollTime = Duration.ofMillis(config.getLong(StreamsConfig.POLL_MS_CONFIG));
+
+        this.changelogs = new HashMap<>();
     }
 
-    @Override
-    public void register(final StateRestorer restorer) {
-        if (!stateRestorers.containsKey(restorer.partition())) {
-            restorer.setUserRestoreListener(userStateRestoreListener);
-            stateRestorers.put(restorer.partition(), restorer);
+    private static String recordEndOffset(final Long endOffset) {
+        return endOffset == null ? "UNKNOWN (since it is for standby task)" : endOffset.toString();
+    }
 
-            log.trace("Added restorer for changelog {}", restorer.partition());
+    private static boolean restoredToEnd(final Long endOffset, final Long currentOffset) {
+        if (endOffset == null) {
+            // end offset is not initialized meaning that it is from a standby task (i.e. it should never end)
+            return false;
+        } else if (endOffset == 0) {
+            // this is a special case, meaning there's nothing to be restored since the changelog has no data
+            return true;
+        } else if (currentOffset == null) {
+            // current offset is not initialized meaning there's no checkpointed offset,
+            // we would start restoring from beginning and it does not end yet
+            return false;
         } else {
-            log.debug("Skip re-adding restorer for changelog {}", restorer.partition());
+            // note the end offset returned of the consumer is actually the last offset + 1
+            return currentOffset + 1 < endOffset;
         }
+    }
 
-        needsInitializing.add(restorer.partition());
+    // Once some new tasks are created, we transit to restore them and pause on the existing standby tasks.
+    // NOTE: even if the newly created tasks do not need any restoring, we still first transit to this state and then
```

Comment:
This is a great simplification: the Streams state, which is based on the task state transitions, should not depend or need to be handled differently based on characteristics of the tasks (eg stateful vs stateless, revoked while still restoring vs running, etc)